silq/analysis/analysis.py

```diff
@@ -715,8 +717,81 @@ def analyse_EPR(empty_traces: np.ndarray,
+
+    Args:
+        up_proportions_arr: 1D Array of up proportions, calculated from
+            `analyse_traces`. Up proportion is the proportion of traces that
+            contain blips. The length is number of samples.
+        shots_per_frequency: Integer number of ESR shots/traces, which are
+            used to determine the up proportion.
+
+    Returns:
+        (float) threshold up-proportion
+    """
+
+    num_shots = max(shots_per_frequency, 25)
+    # Algorithm fails if the number of shots is less than 25 (not exact number, can still be adjusted).
+    # The reason for this is that in the following division of up_proportion space by the number of shots
+    # there are too few points to perform kernel density estimation.
+
+    proportion_space = np.linspace(0, 1, num=num_shots + 1, endpoint=True)
+    up_proportions_arr = up_proportions_arr.reshape(1, -1)    # don't think it's needed
+    kernel = gaussian_kde(up_proportions_arr)
+    gaussian_up_proportions = kernel(proportion_space)
+
+    # Finding indexes of density distribution peaks. Might still require some fine adjustments in
+    # 'thres' and 'min_dist' parameters.
+    peak_idxs = peakutils.peak.indexes(gaussian_up_proportions,
+                                       thres=0.5/up_proportions_arr.shape[-1],
+                                       min_dist=num_shots/5)
+    # If no peaks are detected (something failed), by default we set threshold to 0.5.
+    middle_point = (num_shots + 1) // 2
+    if len(peak_idxs) == 0:
+        logger.debug(f'Adaptive thresholding routine: 0 peaks were '
+                     f'found, using threshold 0.5')
+        trough_idx = middle_point
+    # If only one peak is detected (nuclear spin did not flip), depending on whether
+    # it is in the lower (<0.5) or the upper (>=0.5) half of proportion space,
+    # the minimum index of density distribution is found respectively above or below
+    # this peak's index.
+    # Since the density distribution values reach 1e-20 at the edges of the
+    # proportion space (i.e. at 0 or 1), to smooth a bit the threshold jumps between
+    # the measurement points, rounding to the 4th digit is implemented.
+    elif len(peak_idxs) == 1:
+        if peak_idxs[0] < middle_point:
+            trough_slice = slice(*[peak_idxs[0], num_shots + 1])
+            search_space = np.round(gaussian_up_proportions[trough_slice], 4)
+            trough_idx = peak_idxs[0] + np.argmin(search_space)
+        else:
+            trough_slice = slice(*[0, peak_idxs[0]])
+            search_space = np.round(gaussian_up_proportions[trough_slice], 4)
+            trough_idx = np.argmin(search_space)
+    # If more than one peak is detected, then minimum is found between the lowest and the highest peak.
+    else:
+        trough_slice = slice(*[min(peak_idxs), max(peak_idxs)])
+        search_space = np.round(gaussian_up_proportions[trough_slice], 4)
+        trough_idx = peak_idxs[0] + np.argmin(search_space)
```

Comment:
Why the min? Why not the average?

Reply:
The algorithm aims to find the minimum overlap point between 2 gaussian distributions, hence taking `np.argmin()` of the kernel function, rather than the average of the 2 peaks. 
If the 2 peaks have different widths then the average is not the best choice, however in this application of the algorithm all data is smoothed to the same width I believe. There is room for improvement here.

Reply:
I agree with Mark

Reply:
Hmm maybe add it as a `# TODO {text}` comment?

Reply:
corrected.